File: py/graph/eulerian_path/directed/Fleury.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class EdgeNode:
    def __init__(self, index: int, weight: int):
        self.index, self.weight, self.next = index, weight, None

class VertexNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class Fleury:
    def search(self, graphs: List[VertexNode]) -> List[int]:
        start = self.__getStart(graphs)
        if not ~start:
            return []

        visited, stack = [[False for _ in range(len(graphs))] for _ in range(len(graphs))], [start]
        result = []
        while stack:
            index = stack.pop()
            edge, flag = graphs[index].next, False
            while edge:
                edgeIndex = edge.index
                if not visited[index][edgeIndex]:
                    flag = True
                    break
                edge = edge.next
            if not flag:
                result.append(index)
            else:
                self.__search(graphs, index, stack, visited)
        result.reverse()
        return result

    def __search(self, graphs: List[VertexNode], start: int, stack: List[int], visited: List[List[bool]]) -> None:
        stack.append(start)
        edge = graphs[start].next
        while edge:
            edgeIndex = edge.index
            if not visited[start][edgeIndex]:
                visited[start][edgeIndex] = True
                self.__search(graphs, edgeIndex, stack, visited)
                break
            edge = edge.next

    def __getStart(self, graphs: List[VertexNode]) -> int:
        inDegrees, outDegrees = [0 for _ in range(len(graphs))], [0 for _ in range(len(graphs))]
        for i, value in enumerate(graphs):
            edge = value.next
            while edge:
                inDegrees[edge.index] += 1
                outDegrees[i] += 1
                edge = edge.next

        start, inCount, outCount = 0, 0, 0
        for i, (value1, value2) in enumerate(zip(inDegrees, outDegrees)):
            if abs(value1 - value2) > 1:
                return -1

            if 1 == value2 - value1:
                start = i
                inCount += 1
            elif 1 == value1 - value2:
                outCount += 1
            if inCount > 1 or outCount > 1:
                break
        return start if (not inCount and not outCount) or (1 == inCount and 1 == outCount) else -1

File: cpp/graph/eulerian_path/directed/Fleury.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct EdgeNode {
    int index;
    int weight;
    EdgeNode *next;

    EdgeNode(int index, int weight) : index(index), weight(weight), next(nullptr) {}
};

struct VertexNode {
    int value;
    EdgeNode *next;

    VertexNode(int value) : value(value), next(nullptr) {}
};

class Fleury {
public:
    vector<int> search(vector<VertexNode *> &graphs) {
        int start = getStart(graphs);
        if (!~start) {
            return vector<int>(0);
        }

        vector<vector<bool>> visited(graphs.size(), vector<bool>(graphs.size(), false));
        stack<int> stack;
        stack.push(start);

        vector<int> result(0);
        while (!stack.empty()) {
            int index = stack.top();
            stack.pop();
            EdgeNode *edge = graphs[index]->next;
            bool flag = false;
            while (edge) {
                int edgeIndex = edge->index;
                if (!visited[index][edgeIndex]) {
                    flag = true;
                    break;
                }
                edge = edge->next;
            }
            if (!flag) {
                result.emplace_back(index);
            } else {
                search(graphs, index, stack, visited);
            }
        }
        reverse(result.begin(), result.end());
        return result;
    }

private:
    void search(vector<VertexNode *> &graphs, int start, stack<int> &stack, vector<vector<bool>> &visited) {
        stack.push(start);
        EdgeNode *edge = graphs[start]->next;
        while (edge) {
            int edgeIndex = edge->index;
            if (!visited[start][edgeIndex]) {
                visited[start][edgeIndex] = true;
                search(graphs, edgeIndex, stack, visited);
                break;
            }
            edge = edge->next;
        }
    }

    int getStart(vector<VertexNode *> &graphs) {
        vector<int> inDegrees(graphs.size(), 0);
        vector<int> outDegrees(graphs.size(), 0);
        for (int i = 0; i < (int) graphs.size(); ++i) {
            EdgeNode *edge = graphs[i]->next;
            while (edge) {
                ++inDegrees[edge->index];
                ++outDegrees[i];
                edge = edge->next;
            }
        }

        int start = 0;
        int inCount = 0;
        int outCount = 0;
        for (int i = graphs.size() - 1; i >= 0; --i) {
            if (abs(inDegrees[i] - outDegrees[i]) > 1) {
                return -1;
            }

            if (1 == outDegrees[i] - inDegrees[i]) {
                start = i;
                ++inCount;
            } else if (1 == inDegrees[i] - outDegrees[i]) {
                ++outCount;
            }
            if (inCount > 1 || outCount > 1) {
                break;
            }
        }
        return ((!inCount && !outCount) || (1 == inCount && 1 == outCount)) ? start : -1;
    }
};
